system/core/tags.php

```diff
@@ -0,0 +1,184 @@
+<?php
+	// TODO:
+	// - add styling for default theme
+	// - implement error notification if new tag exists
+	// - implement tag management page: edit, translate, remove or merge tags
+	// - implement tagcontainer pagetype
+	
+	function tagList($pageListNode, $lang) {
+		global $hyphaXml;
+		
+		// Build list of active tags for given page
+		foreach($pageListNode->getElementsByTagName('tag') as $tag) $activeTags[] = $tag->getAttribute('id');
+		
+		// Divide list of all tags stored in hypha.xml in a selected tag list and a deselected tag list
+		foreach($hyphaXml->getElementsByTagName('tag') as $tag) {
+			foreach($tag->getElementsByTagName('language') as $tagLang) {
+				if ($tagLang->getAttribute('id')==$lang) {
+					if (!empty($activeTags) && in_array($tag->getAttribute('id'), $activeTags)) $selectedTags[$tag->getAttribute('id')] = array("label" => $tagLang->getAttribute('label'), "description" => $tagLang->nodeValue);
+					else $deselectedTags[$tag->getAttribute('id')] = array("label" => $tagLang->getAttribute('label'), "description" => $tagLang->nodeValue);
+				}
+			}
+		}
+		
+		// Construct html output
+		$html = '';
+		if (!empty($activeTags) || isUser()) {
+			$html.= '<span class="prefix">' . __('tags') . ': </span>';
+			$html.= '<div id="selectedTags" style="display:inline-block;">';
+			if (!empty($selectedTags)) foreach ($selectedTags as $id => $tag) {
+				$html.= '<div id="tagSel_'.$id.'" class="tag">';
+				$html.= '<span title="'.$tag['description'].'">'.$tag['label'].'</span>';
+				if (isUser()) $html.= ' <span style="color:#8f0000; cursor:pointer;" title="'.__('remove-tag').'" onclick="deselectTag(\''.$id.'\');">⨯</span>';
+				$html.= '</div>';
+			}
+			$html.= '</div>';
+			if ((isUser() && !empty($deselectedTags)) || isAdmin()) {
+				$html.= '<select id="remainingTags" style="padding:0px;" onchange="selectTag(this.value);">';
+				$html.= '<option value="_add_" selected="selected" disabled="disabled">'.__('add').'...</option>';
+				if (!empty($deselectedTags)) foreach ($deselectedTags as $id => $tag) {
+					$html.= '<option id="tagRem_'.$id.'" value="'.$id.'">'.$tag['label'].'</option>';
+				}
+				if (isAdmin()) $html.= '<option value="">'.__('new').'...</option>';
+				$html.= '</select>';
+			}
+		}
+		return $html;
+	}
+	
+	/*
+		Function: hypha_indexTags
+		returns list of tags for the given page
+
+		Parameters:
+		$pageListNode - DOMElement containing page settings
+		$lang - language for tag index
+	*/
+	function hypha_indexTags($pageListNode, $lang) {
+		global $hyphaHtml;
+		
+		if (!$pageListNode || !$lang) return '';
+		
+		$language = hypha_pageGetLanguage($pageListNode, $lang);
+		$url = $language->getAttribute('id').'/'.$language->getAttribute('name');
+		
+		ob_start();
+?>
+<script>
+	function selectTag(id) {
+		if (!id) addTag();
+		else {
+			hypha('<?=$url?>', 'tagSelect', id, function(response) {document.getElementById('tagList').innerHTML = response;});
```

Comment:
Again, I would use `json_encode` and `htmlspecialchars` here and in similar places below.

Reply:
Fixed.

Reply:
I ended up not using `htmlspecialchars` here, since HTML entities are not interpreted inside `<script>` tags. This does leave a small window for problems - if a string value contains `</script>` (or even `</` perhaps), then this could end this script tag. This seems unlikely to happen in practice, though it might be a small security risk. For now, let's just leave it like this (this could perhaps be handled more thoroughly by generating HTML through DomDocument everywhere later).